docs/Configuration.md
# SituationBoard Configuration
The configuration of SituationBoard is stored in the ```situationboard.conf``` file.
It is necessary to adjust these settings before starting SituationBoard for the first time.
The configuration file is divided into several sections.
Within each section individual settings are specified in form of key/value pairs:
```
[section]
key = value
# some comment
```

In addition to the settings stated below, all source and action plugins also feature a ```debug``` setting that can be used (```debug = True```) to enable plugin specific debug messages.

Important and frequently used settings are highlighted in **bold**.

## General Settings

### Backend
The backend settings adjust the general behaviour of the server backend.
Currently, the following settings are available in the ```[backend]``` section of the configuration:

| Setting                      | Description                                                                                                                                                       | Default Value                 |
|------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------|
| **server_host**              | Hostname or IP address for the server to listen on (127.0.0.1: available only locally; 0.0.0.0: available also for other machines)                                | "127.0.0.1" (local only)      |
| **server_port**              | Port number for the server to listen on                                                                                                                           | 5000                          |
| debug                        | Enable debug mode for the backend server                                                                                                                          | False                         |
| reloader                     | Enable Flask reloader                                                                                                                                             | False                         |
| web_api                      | Enable Web API to allow retrieval of system state and statistics                                                                                                  | False                         |
| loop_sleep_duration          | Duration slept in the event loop between polling/handling of alarm events (in seconds)                                                                            | 1 second                      |
| **sources**                  | Comma-separated list of active sources (i.e. source plugins)                                                                                                      | [] (none)                     |
| **actions**                  | Comma-separated list of active actions (i.e. action plugins)                                                                                                      | [] (none)                     |

### Frontend
The frontend settings adjust the appearance and behaviour of the standby view and the alarm view.
Currently, the following settings are available in the ```[frontend]``` section of the configuration:

| Setting                      | Description                                                                                                                                                       | Default Value                 |
|------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------|
| **header**                   | Header of the standby view (typically the name of the organization)                                                                                               | "Feuerwehr Musterdorf"        |
| **news**                     | Important news shown in the standby view (leave empty to hide this line)                                                                                          | "Herzlich Willkommen"         |
| debug                        | Enable debug output for frontend (java script)                                                                                                                    | False                         |
| **language**                 | Language of the frontend UI; currently only english (en) and german (de) is available                                                                             | "en"                          |
| **alarm_duration**           | Duration until an alarm disappears and the standby view becomes active again (in seconds)                                                                         | 60 * 60 seconds (1 hour)      |
| **alarm_show_maps**          | Maps shown during an alarm: Alarm Location ("location"); Route ("route"); "both"; "none"                                                                          | "both"                        |
| standby_show_statistics      | Enable statistics in the standby view                                                                                                                             | True                          |
| standby_show_clock           | Enable clock (and date) in the standby view                                                                                                                       | True                          |
| page_reload_duration         | Duration between two page reloads of the frontend (in seconds; 0 = never)                                                                                         | 0 (never)                     |
| calendar_url                 | URL of the calendar (ICS) file – should typically match the destination_url setting of the update_calendar plugin but exclude the "frontend/" prefix              | "data/calendar.ics"           |
| calendar_update_duration     | Duration between two updates of the calendar in the frontend (in seconds; 0 = never) – set this to never (0) if you use the update_calendar plugin                | 0 (never)                     |
| map_service                  | The map service used: Google Maps ("google") or Open Street Maps ("osm")                                                                                          | "osm"                         |
| map_api_key                  | Your personal API key (only required for Google Maps API v3)                                                                                                      | "" (none)                     |
| map_zoom                     | Map zoom level 1.0 - 20.0 depending on the map service (only affects location map; route map will show the whole route)                                           | 19.0                          |
| map_type                     | Different map views ("default", "roadmap", "satellite", "hybrid", "terrain") of the map (currently only supported by Google Maps)                                 | "default" (= "hybrid")        |
| map_emergency_layer          | Show emergency objects ("none", "fire", "medical", "all") in the map (currently only supported by OSM)                                                            | "all"                         |
| **map_home_latitude**        | Latitude location of the fire/emergency station (i.e. start of the journey for map view)                                                                          | 0.0                           |
| **map_home_longitude**       | Longitude location of the fire/emergency station (i.e. start of the journey for map view)                                                                         | 0.0                           |
| **map_search_location**      | Determines whether the frontend tries to search the location based on the address if there are no coordinates of the location provided in the alarm event         | False                         |
| show_splash_screen           | Determines whether a splash screen is shown when the frontend is loaded                                                                                           | True                          |

## Sources
Alarm sources allow the detection of alarm events.
They are activated by adding them to the list of ```sources``` in the ```[backend]``` section of the configuration:
```
sources = source1,source2,...

# for example:
sources = dummy,binary,sms
```
Each source has its own configuration section:
```
# for example:
[source:binary]
key = value
```
Some source plugins may support multiple instances.
When using a source plugin more than once, it is required to append a unique instance name
to the individual source names (separated by ```:```):
```
# for example:
sources = binary:one,binary:two,sms
```
The same holds for the corresponding configuration sections:
```
# for example:
[source:binary:one]
key = value

[source:binary:two]
key = value
```

Sources are scanned in the order they are specified in the list of sources.
Currently, the following sources are available:

### SMS Source
The ```sms``` source allows to receive alarm events via SMS messages.
The SMS messages are received with the help of the [Gammu](https://wammu.eu/gammu/) SMS/phone library and then parsed
with a specified parser to retrieve information on the individual alarms (like the actual alarm event or location).
The SMS source is configured in the ```[source:sms]``` section and has the following settings:

| Setting                      | Description                                                                                                                                                       | Default Value                 |
|------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------|
| gammu_config                 | Path to the Gammu configuration file                                                                                                                              | "/etc/gammu-smsdrc"           |
| **parser**                   | Selected message parser                                                                                                                                           | "sms"                         |
| allowlist                    | Comma-separated list of allowed/accepted phone numbers in E.164 / FQTN format without whitespace/slashes (e.g. +491234567890)                                     | [] (accept all)               |
| denylist                     | Comma-separated list of denied/rejected phone numbers in E.164 / FQTN format without whitespace/slashes (e.g. +491234567890)                                      | [] (reject none)              |

Depending on your setup, further adjustments (e.g. modem device, modem connection or SIM PIN)
might be necessary in the Gammu configuration file ```/etc/gammu-smsdrc```.

#### Tested Cellular USB Modems
The following cellular USB modems were tested successfully with the stated configuration parameters:

| Cellular USB Modem          | USB-ID (as modem)  | Device        | Connection  | Other USB-IDs (used in other operation modes)         |
|-----------------------------|--------------------|---------------|-------------|-------------------------------------------------------|
| **Huawei E303 (51077742)**  | 12d1:1506          | /dev/ttyUSB0  | at          | (none)                                                |
| **Huawei E303 (51070NQP)**  | 12d1:1001          | /dev/ttyUSB0  | at          | 12d1:1f01 (storage device), 12d1:14dc (network card)  |

The devices in this table are identified by a USB-ID consisting of a vendor ID and a product ID.
Some USB devices can operate in different modes (e.g. as a storage device containing device drivers,
a network card and as an actual modem). These different modes are often represented by different product IDs:

| Cellular USB Modem          | Vendor ID | Vendor                         | Product ID | Product                                      | Device Mode    | Comment                                                  |
|-----------------------------|-----------|--------------------------------|------------|----------------------------------------------|----------------|----------------------------------------------------------|
| **Huawei E303 (51077742)**  | **12d1**  | Huawei Technologies Co., Ltd.  | **1506**   | Modem/Networkcard                            | **modem**      | required mode (default and only mode)                    |
| **Huawei E303 (51070NQP)**  | **12d1**  | Huawei Technologies Co., Ltd.  | 1f01       | E353/E3131 (Mass storage mode)               | storage device | default mode (without usb_modeswitch)                    |
|                             |           |                                | 14dc       | E33372 LTE/UMTS/GSM HiLink Modem/Networkcard | network card   | default target mode (with stock usb_modeswitch)          |
|                             |           |                                | **1001**   | E161/E169/E620/E800 HSDPA Modem              | **modem**      | required mode (enforced with custom usb_modeswitch file) |

You can retrieve the (current) vendor/product ID of your USB modem with the ```lsusb``` command line tool.
The ```usb_modeswitch``` tool makes sure that the device is in the correct mode for usage as a modem.
After the SituationBoard installation, a reboot might be necessary before your device is recognized correctly.

**Important:**
USB modems can consume a lot of power (often exceeding the capabilities of the Raspberry Pi).
Therefore, it often makes sense to operate them on an externally-powered USB hub.

#### SMS Message Parser
Currently, there is only one message parser (the default ```sms``` parser) available.
It is configured in the ```[parser:sms]``` section and has the following settings:

| Setting                      | Description                                                                                                                                                       | Default Value                 |
|------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------|
| merge_duration               | Duration in which non-standard multipart SMS are merged (in seconds)                                                                                              | 90 seconds                    |
| **alarm_header**             | Message header identifying an alarm message                                                                                                                       | "SMS Alarm"                   |
| **alarm_senders**            | Comma-separated list of typical senders of alarms (phone number in E.164 / FQTN format) whose messages will be treated as alarm even if the header does not match | [] (none)                     |

Regarding alarm events, the SMS parser supports the following message format:
```
<ALARM_HEADER>
Alarmzeit: <DD.MM.YYYY HH:MM:SS>
EO: <LOCATION>, <LOCATION_DETAILS>
<EVENT_DETAILS>
STW: <EVENT>
Bem: <MULTILINE_COMMENT>
<MULTILINE_COMMENT>
<MULTILINE_COMMENT>
```

An example SMS could therefore look like:
```
ILS BA-FO SMS Alarm
Alarmzeit: 03.09.2020 16:24:12
EO: Musterort - Mustergemeinde, Musterstraße 1
Technische Hilfe / #T2719#klein#Straße reinigen
STW: THL 1
Bem: Längere Ölspur
Beginn kurz vor Aussiedlerhof
```

### Divera Source
The `divera` source allows to retrieve alarms from the web API of the [Divera 24/7](https://www.divera247.com) alarm service.
Alongside the actual alarm infos it is also possible to retrieve the responses of crew members to an alarm and the current status of vehicles.
The Divera source is configured in the ```[source:divera]``` section and has the following settings:

| Setting                      | Description                                                                                                                                                       | Default Value                 |
|------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------|
| **api_key**                  | API key (also called access key by Divera) of the organization (required)                                                                                         | ""                            |
| timeout                      | Request timeout (in seconds)                                                                                                                                      | 5 seconds                     |
| api_url                      | URL of the API (adjust only when testing with a manual mock API setup)                                                                                            | "https://www.divera247.com"   |
| use_mock_api                 | Use mock API instead of the specified API URL to simulate alarm events (starts a local mock API server; ignores the specified api_url and api_key)                | False                         |
| ignore_test_alarm            | Ignore test alarms (i.e. alarms with an author_id of 0)                                                                                                           | False                         |
| **show_vehicle_status**      | Update the news setting with the current status of the vehicles to show it in the standby view                                                                    | True                          |
| **show_crew_responses**      | Append the crew responses to the comment of the alarm event to show them in the alarm view                                                                        | True                          |
| response_id_fast             | Identifier used for responses that signal fast availability of a crew member (< 5 min)                                                                            | "99148"                       |
| response_id_slow             | Identifier used for responses that signal slow availability of a crew member (< 10 min)                                                                           | "99149"                       |
| response_id_na               | Identifier used for responses that signal no availability of a crew member (n/a)                                                                                  | "99150"                       |

### Binary Source
The ```binary``` source allows to detect alarm events based on a GPIO input of the Raspberry Pi.
A typical use case for the ```binary``` source is to connect a pager as alarm source (e.g. as fallback or to forward alarm events to emergency personnel).
Unlike the SMS source it does not provide additional information on the alarm (like the actual alarm event or location).
The binary source is configured in the ```[source:binary]``` section and has the following settings:

| Setting                      | Description                                                                                                                                                       | Default Value                 |
|------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------|
| **message**                  | Text show as alarm message                                                                                                                                        | "Alarm"                       |
| **pin**                      | BCM pin of the Raspberry Pi used for alarm detection                                                                                                              | 13 (BCM Pin 13)               |
| **active_high**              | Determines whether the IO pin is active high                                                                                                                      | False                         |

By appending unique instance names (e.g. ```binary:pager``` and ```[source:binary:pager]```),
more than one ```binary``` source can be used at the same time.

### Dummy Source
The ```dummy``` source allows to debug the system and its configuration without the need to trigger alarm events via SMS or binary input.
It generates dummy messages when a specific signal is sent to the backend process (e.g. via the ```sbctl``` tool).
Currently, the dummy source offers no settings.

To simulate a regular (text) alarm use:
```
sbctl dalarm
```

To simulate a binary alarm use:
```
sbctl dbinary
```

To simulate a remote update to settings (for example to the header or news line) use:
```
sbctl dsetting
```

## Actions
Actions are triggered by alarm events or on a regular timely manner.
They are activated by adding them to the list of ```actions``` in the ```[backend]``` section of the configuration:
```
actions = action1,action2,...

# for example:
actions = search_location,update_database,update_settings,update_frontend,update_calendar,send_poweralarm,toggle_outlet,activate_screen
```
Each action has its own configuration section:
```
# for example:
[action:toggle_outlet]
key = value
```
Some action plugins may support multiple instances.
When using an action plugin more than once, it is required to append a unique instance name
to the individual action names (separated by ```:```):
```
# for example:
actions = ...,toggle_outlet:one,toggle_outlet:two,...
```
The same holds for the corresponding configuration sections:
```
# for example:
[action:toggle_outlet:one]
key = value

[action:toggle_outlet:two]
key = value
```

Actions are triggered in the order they are specified in the list of actions.
Currently, the following actions are available:

### Search Location Action (should be 1st)
The ```search_location``` action is responsible to find the location (longitude and latitude) of an alarm event when only the address is known.
It should always be the first action since it might add location information to the alarm event that is useful for the following actions.
It is configured in the ```[action:search_location]``` section and has the following settings:

| Setting                      | Description                                                                                                                                                       | Default Value                 |
|------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------|

CHANGELOG.md
# Changelog

This changelog lists the changes of all the SituationBoard versions.

Breaking changes (e.g. regarding configuration, database and CSV) always come with a version bump.
In contrast, minor changes/bugfixes might not come with a new version number.

## Version 1.15 (2025-10-01)

- Add source driver for the Divera 24/7 API
- Correct handling of updated and outdated alarm events
- Update GPIO support to work with new Raspberry Pi versions
- Update various dependencies and make related adjustments
- Improve automatic screen activation and CEC support
- Make storing of setting updates optional
- Make location search in frontend optional
- Add support for Wayland
- Adjust features installed by default
- Update and improve documentation

## Version 1.14 (2023-08-21)

- Fix websocket bug in backend
- Update frontend dependencies (OpenLayers)
- Replace xorg.conf with xset call at runtime

## Version 1.13 (2022-09-27)

- Introduce a generic list view widget
- Add support for repeating events and all-day events to the calendar widget
- Log source state changes and improve overall log output
- Specify required versions for Python and JavaScript dependencies
- Minor code cleanup and minor other improvements

## Version 1.12 (2021-02-23)

First public release


backend/util/Settings.py
# SituationBoard - Alarm Display for Fire Departments
# Copyright (C) 2017-2021 Sebastian Maier
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

import os
import configparser

from typing import List

from backend.util.Module import Module
from backend.util.StringConverter import StringConverter

class Settings(Module):
    """The Settings class is responsible for reading/writing of user settings from/to the configuration file.
    It is mainly responsible for backend and frontend settings, but also provides the infrastructure which allows
    modules and plugins to have additional settings in their own sections of the configuration file."""

    # Sections
    SECTION_BACKEND                = "backend"
    SECTION_FRONTEND               = "frontend"

    # Keys of writable settings
    KEY_FRONTEND_HEADER            = "header"
    KEY_FRONTEND_NEWS              = "news"

    # Defaults of writable settings
    DEFAULT_FRONTEND_HEADER        = "Feuerwehr Musterdorf"  # organisation shown as header in standby mode
    DEFAULT_FRONTEND_NEWS          = "Herzlich Willkommen"   # news shown as message below header

    def __init__(self, configPath: str, appPath: str, forceDebug: bool = False) -> None:
        super().__init__("settings", settings = None, debug = False)
        self.__filename = configPath
        self.__appPath = appPath
        self.__forceDebug = forceDebug
        self.__config = configparser.RawConfigParser()

        self.print("Reading settings from config file")
        try:
            with open(self.__filename, 'r') as configFile:
                self.__config.read_file(configFile, self.__filename)
        except Exception:
            self.fatal(f"Could not read config file ({self.__filename})")

    def store(self) -> None:
        self.print("Writing settings to config file")
        try:
            with open(self.__filename, 'w') as configFile:
                self.__config.write(configFile)
        except Exception:
            self.error(f"Could not write config file ({self.__filename}")

    def getString(self, section: str, key: str, default: str) -> str:
        try:
            singleLineValue = self.__config.get(section, key)
            value = StringConverter.singleline2string(singleLineValue)
        except Exception:
            value = default

        self.dbgPrint(f"{section}/{key} = {value} (default = {default})")
        return value

    def getBoolean(self, section: str, key: str, default: bool) -> bool:
        try:
            value = self.__config.getboolean(section, key)
        except Exception:
            value = default

        self.dbgPrint(f"{section}/{key} = {value} (default = {default})")
        return value

    def getFloat(self, section: str, key: str, default: float) -> float:
        try:
            value = self.__config.getfloat(section, key)
        except Exception:
            value = default

        self.dbgPrint(f"{section}/{key} = {value} (default = {default})")
        return value

    def getInt(self, section: str, key: str, default: int) -> int:
        try:
            value = self.__config.getint(section, key)
        except Exception:
            value = default

        self.dbgPrint(f"{section}/{key} = {value} (default = {default})")
        return value

    def getList(self, section: str, key: str, default: List[str]) -> List[str]:
        try:
            text = self.__config.get(section, key)
            value = [x.strip() for x in text.split(',') if x.strip() != ""]
        except Exception:
            value = default

        if self.isDebug():
            strValue = ','.join(map(str, value))
            strDefault = ','.join(map(str, default))
            self.dbgPrint(f"{section}/{key} = [{strValue}] (default = [{strDefault}])")

        return value

    def getFilename(self, section: str, key: str, default: str) -> str:
        try:
            filename = self.__config.get(section, key)
        except Exception:
            filename = default

        # filename = filename.replace("$SBPATH", self.__appPath)
        if not os.path.isabs(filename):
            filename = os.path.join(self.__appPath, filename)

        self.dbgPrint(f"{section}/{key} = {filename} (default = {default})")
        return filename

    def getOption(self, section: str, key: str, options: List[str], default: str) -> str:
        try:
            option = self.__config.get(section, key)
            if option not in options:
                option = default
        except Exception:
            option = default

        self.dbgPrint(f"{section}/{key} = {option} (default = {default})")
        return option

    def __setSetting(self, section: str, key: str, value: str) -> bool:
        try:
            if not self.__config.has_section(section):
                self.__config.add_section(section)

            self.__config.set(section, key, value)
            self.print(f"Updated setting {section}/{key} with value {value}")
            return True
        except Exception:
            self.error(f"Failed to update setting {section}/{key} with value {value}")
            return False

    def setString(self, section: str, key: str, value: str) -> bool:
        singleLineValue = StringConverter.string2singleline(value)
        return self.__setSetting(section, key, singleLineValue)

    def setBoolean(self, section: str, key: str, value: bool) -> bool:
        return self.__setSetting(section, key, str(value))

    def setFloat(self, section: str, key: str, value: float) -> bool:
        return self.__setSetting(section, key, str(value))

    def setInt(self, section: str, key: str, value: int) -> bool:
        return self.__setSetting(section, key, str(value))

    def getForceDebug(self) -> bool:
        return self.__forceDebug

    # Backend Settings

    def getBackendServerHost(self) -> str:
        return self.getString(Settings.SECTION_BACKEND, "server_host", "127.0.0.1")

    def getBackendServerPort(self) -> int:
        return self.getInt(Settings.SECTION_BACKEND, "server_port", 5000)

    def getBackendDebug(self) -> bool:
        if self.getForceDebug():
            return True

        return self.getBoolean(Settings.SECTION_BACKEND, "debug", False)

    def getBackendReloader(self) -> bool:
        return self.getBoolean(Settings.SECTION_BACKEND, "reloader", False)

    def getBackendWebAPI(self) -> bool:
        return self.getBoolean(Settings.SECTION_BACKEND, "web_api", False)

    def getBackendLoopSleepDuration(self) -> int:
        return self.getInt(Settings.SECTION_BACKEND, "loop_sleep_duration", 1) # in seconds

    def getBackendSources(self) -> List[str]:
        return self.getList(Settings.SECTION_BACKEND, "sources", [])

    def getBackendActions(self) -> List[str]:
        return self.getList(Settings.SECTION_BACKEND, "actions", [])

    # Frontend Settings

    def getFrontendHeader(self) -> str:
        return self.getString(Settings.SECTION_FRONTEND, Settings.KEY_FRONTEND_HEADER, Settings.DEFAULT_FRONTEND_HEADER)

    def setFrontendHeader(self, value: str) -> bool:
        return self.setString(Settings.SECTION_FRONTEND, Settings.KEY_FRONTEND_HEADER, value)

    def getFrontendNews(self) -> str:
        return self.getString(Settings.SECTION_FRONTEND, Settings.KEY_FRONTEND_NEWS, Settings.DEFAULT_FRONTEND_NEWS)

    def setFrontendNews(self, value: str) -> bool:
        return self.setString(Settings.SECTION_FRONTEND, Settings.KEY_FRONTEND_NEWS, value)

    def getFrontendDebug(self) -> bool:
        return self.getBoolean(Settings.SECTION_FRONTEND, "debug", False)

    def getFrontendLanguage(self) -> str:
        availableLanguages = ["en", "en_us", "en_gb", "de"]
        defaultLanguage = "en"
        return self.getOption(Settings.SECTION_FRONTEND, "language", availableLanguages, defaultLanguage)

    def getFrontendAlarmDuration(self) -> int:
        return self.getInt(Settings.SECTION_FRONTEND, "alarm_duration", 60 * 60) # in seconds - default 1 hour

    def getFrontendAlarmShowMaps(self) -> str:
        availableMaps = ["none", "location", "route", "both"]
        defaultMaps = "both"
        return self.getOption(Settings.SECTION_FRONTEND, "alarm_show_maps", availableMaps, defaultMaps)

    def getFrontendCalendarURL(self) -> str:
        return self.getString(Settings.SECTION_FRONTEND, "calendar_url", "data/calendar.ics")

    def getFrontendCalendarUpdateDuration(self) -> int:
        return self.getInt(Settings.SECTION_FRONTEND, "calendar_update_duration", 0) # in seconds - default 0 = never

    def getFrontendStandbyShowStatistics(self) -> bool:
        return self.getBoolean(Settings.SECTION_FRONTEND, "standby_show_statistics", True)

    def getFrontendStandbyShowClock(self) -> bool:
        return self.getBoolean(Settings.SECTION_FRONTEND, "standby_show_clock", True)

    def getFrontendPageReloadDuration(self) -> int:
        return self.getInt(Settings.SECTION_FRONTEND, "page_reload_duration", 0) # in seconds - default 0 = never

    def getFrontendMapService(self) -> str:
        avaiableServices = ["osm", "osm_openlayers", "osm_leaflet", "google"]
        defaultService = "osm"
        return self.getOption(Settings.SECTION_FRONTEND, "map_service", avaiableServices, defaultService)

    def getFrontendMapAPIKey(self) -> str:
        return self.getString(Settings.SECTION_FRONTEND, "map_api_key", "")

    def getFrontendMapZoom(self) -> float:
        return self.getFloat(Settings.SECTION_FRONTEND, "map_zoom", 19.0)

    def getFrontendMapType(self) -> str:
        availableTypes = ["default", "roadmap", "satellite", "hybrid", "terrain"]
        defaultType = "default"
        return self.getOption(Settings.SECTION_FRONTEND, "map_type", availableTypes, defaultType)

    def getFrontendMapEmergencyLayer(self) -> str:
        availableLayers = ["none", "medical", "fire", "all"]
        defaultLayer = "all"
        return self.getOption(Settings.SECTION_FRONTEND, "map_emergency_layer", availableLayers, defaultLayer)

    def getFrontendMapHomeLatitude(self) -> float:
        return self.getFloat(Settings.SECTION_FRONTEND, "map_home_latitude", 0.0)

    def getFrontendMapHomeLongitude(self) -> float:
        return self.getFloat(Settings.SECTION_FRONTEND, "map_home_longitude", 0.0)

    def getFrontendMapSearchLocation(self) -> bool:
        return self.getBoolean(Settings.SECTION_FRONTEND, "map_search_location", False)

    def getFrontendShowSplashScreen(self) -> bool:
        return self.getBoolean(Settings.SECTION_FRONTEND, "show_splash_screen", True)
